tests/core_tests/loki_tests.cpp

```diff
@@ -701,6 +701,74 @@ bool loki_core_governance_batched_reward::generate(std::vector<test_event_entry>
   return true;
 }
 
+bool loki_core_block_rewards_lrc6::generate(std::vector<test_event_entry>& events)
+{
+  auto& network = cryptonote::get_config(cryptonote::FAKECHAIN, cryptonote::network_version_16);
+  std::vector<std::pair<uint8_t, uint64_t>> hard_forks = loki_generate_sequential_hard_fork_table(cryptonote::network_version_15_lns);
+  hard_forks.emplace_back(cryptonote::network_version_16, hard_forks.back().second + network.GOVERNANCE_REWARD_INTERVAL_IN_BLOCKS + 10);
+  loki_chain_generator batched_governance_generator(events, hard_forks);
+  batched_governance_generator.add_blocks_until_version(cryptonote::network_version_16);
+  batched_governance_generator.add_n_blocks(network.GOVERNANCE_REWARD_INTERVAL_IN_BLOCKS);
+
+  uint64_t hf15_height = 0, hf16_height = 0;
+  for (const auto &hf : hard_forks)
+  {
+    if (hf.first == cryptonote::network_version_15_lns)
+      hf15_height = hf.second;
+    else if (hf.first == cryptonote::network_version_16)
+    {
+      hf16_height = hf.second;
+      break;
+    }
+  }
+
+  loki_register_callback(events, "check_lrc6_block_rewards", [hf15_height, hf16_height, interval=network.GOVERNANCE_REWARD_INTERVAL_IN_BLOCKS](cryptonote::core &c, size_t ev_index)
+  {
+    DEFINE_TESTS_ERROR_CONTEXT("check_lrc6_block_rewards");
+
+    uint64_t height = c.get_current_blockchain_height();
+    std::vector<cryptonote::block> blockchain;
+    if (!c.get_blocks((uint64_t)0, (size_t)height, blockchain))
+      return false;
+
+    int hf15_gov = 0, hf16_gov = 0;
+    for (size_t block_height = hf15_height; block_height < hf16_height; ++block_height)
+    {
+      const cryptonote::block &block = blockchain[block_height];
+      CHECK_EQ(block.miner_tx.vout.at(0).amount, MINER_REWARD_HF15);
+      CHECK_EQ(block.miner_tx.vout.at(1).amount, SN_REWARD_HF15);
+      if (cryptonote::block_has_governance_output(cryptonote::FAKECHAIN, block))
+      {
+        hf15_gov++;
+        CHECK_EQ(block.miner_tx.vout.at(2).amount, FOUNDATION_REWARD_HF15 * interval);
+        CHECK_EQ(block.miner_tx.vout.size(), 3);
+      }
+      else
+        CHECK_EQ(block.miner_tx.vout.size(), 2);
+    }
+    for (size_t block_height = hf16_height; block_height < height; ++block_height)
+    {
+      const cryptonote::block &block = blockchain[block_height];
+      CHECK_EQ(block.miner_tx.vout.at(0).amount, 0);
```

Comment:
This is causing the crash in the tests.

At HF16, if we remove the miner reward, we don't need to generate an output for miners anymore in construct_miner_tx. The testing suite assumes that if there's a 0 amount output (as it currently is), it's an RCT transaction and attempts to decode the output and dereferences a nullptr in the ring signatures causing the crash. 

Removing the miner output might be a bit tricky in that you have to find all the code that assumes the miner reward and service node reward are in index 0 and 1 respectively and adjust them accordingly.

Reply:
We still need to include a miner output in the tx because tx fees still have to go somewhere, and we still need block size penalties to allow the block scaling to work properly.

But it's something to be solved later, for now I'll just comment out the hf16 part of the test.

Reply:
Oh, I see, it's not that simple.  *This* test is fine, it's elsewhere in the test suite that has trouble.

Reply:
Easiest might be to just remove HF16, or move it to after the count and change this test to only go up to network type count-1. Most of the other core tests generate hard forks up to count-1

Reply:
> We still need to include a miner output in the tx because tx fees still have to go somewhere

Though of course we can drop it when there are no tx fees.

To hack around this for now I've added +1 atomic loki to the HF16 fee to be the HF16 miner fee until we deal with it.